File: Lua/Operators/OperatorPrecedence.lua
-- Operator Precedence

-- Variables for demonstration
local a = 5
local b = 10
local c = 20

-- 1. Arithmetic Operators
local result1 = a + b * c
print("Result of a + b * c (b * c is evaluated first):", result1)  -- Output: 205

-- Use parentheses to change precedence
local result2 = (a + b) * c
print("Result of (a + b) * c (a + b is evaluated first):", result2)  -- Output: 300

-- 2. Relational Operators
local result3 = a < b and b < c
print("Result of a < b and b < c (relational before logical):", result3)  -- Output: true

-- 3. Logical Operators
local result4 = not a < b  -- 'not' has higher precedence than '<'
print("Result of not a < b (not is evaluated first):", result4)  -- Output: false

-- Use parentheses to change precedence
local result5 = not (a < b)
print("Result of not (a < b) (a < b is evaluated first):", result5)  -- Output: false

-- 4. Concatenation Operator
local str1 = "Hello"
local str2 = "World"
local result6 = str1 .. " " .. str2
print("Result of str1 .. ' ' .. str2 (concatenation from left to right):", result6)  -- Output: Hello World

-- 5. Mixing Different Operators
local result7 = a + b > c and not (a > b)
print("Result of a + b > c and not (a > b):", result7)  -- Output: false

-- 6. Length Operator
local lengthStr = #"Hello"
local result8 = lengthStr + a
print("Result of #\"Hello\" + a (length evaluated first):", result8)  -- Output: 10

-- Combining everything to show precedence
local complexResult = (a + b) * c / (a - b) + a^2 - b % a
print("Complex expression ((a + b) * c / (a - b) + a^2 - b % a):", complexResult)

-- Explanation of complex expression
-- ((a + b) * c / (a - b) + a^2 - b % a)
-- 1. (a + b) -> 15
-- 2. (a - b) -> -5
-- 3. a^2 -> 25
-- 4. b % a -> 0
-- 5. (a + b) * c -> 300
-- 6. 300 / (a - b) -> 300 / -5 -> -60
-- 7. -60 + a^2 -> -60 + 25 -> -35
-- 8. -35 - b % a -> -35 - 0 -> -35
print("Explanation: ((a + b) * c / (a - b) + a^2 - b % a) ->", complexResult)

File: C++/Basic Examples/Operators/OperatorPrecedence.cpp
#include <iostream>
using namespace std;

int main() {
    // Operator Precedence Example

    // Arithmetic Operators
    int result1 = 5 + 3 * 2;  // '*' has higher precedence than '+'
    int result2 = (5 + 3) * 2;  // Parentheses () have the highest precedence
    int result3 = 10 / 2 + 3;  // '/' has higher precedence than '+'
    int result4 = 10 / (2 + 3);  // Parentheses () can be used to change precedence

    // Relational Operators
    bool isTrue = (result1 > result2) && (result3 <= result4);  // '>' and '<=' have higher precedence than '&&'

    // Logical Operators
    bool logic1 = true || false && !isTrue;  // '!' has the highest precedence, followed by '&&', then '||'

    // Bitwise Operators
    int bitwise1 = 5 & 3 | 1;  // '&' has higher precedence than '|'

    // Assignment Operators
    int a = 10;
    a += 5 * 2;  // '*=' has higher precedence than '+='

    // Conditional Operator (Ternary)
    int x = 10, y = 20;
    int max = (x > y) ? x : y;  // '?' and ':' have lower precedence than arithmetic and relational operators

    // Comma Operator
    int b = (x++, y++);  // ',' has the lowest precedence

    // Output results
    cout << "Arithmetic Operators:" << endl;
    cout << "5 + 3 * 2 = " << result1 << endl;  // Expected: 11
    cout << "(5 + 3) * 2 = " << result2 << endl;  // Expected: 16
    cout << "10 / 2 + 3 = " << result3 << endl;  // Expected: 8
    cout << "10 / (2 + 3) = " << result4 << endl;  // Expected: 2
    cout << endl;

    cout << "Relational Operators:" << endl;
    cout << "result1 > result2 && result3 <= result4 = " << isTrue << endl;  // Expected: 1 (true)
    cout << endl;

    cout << "Logical Operators:" << endl;
    cout << "true || false && !isTrue = " << logic1 << endl;  // Expected: 1 (true)
    cout << endl;

    cout << "Bitwise Operators:" << endl;
    cout << "5 & 3 | 1 = " << bitwise1 << endl;  // Expected: 5
    cout << endl;

    cout << "Assignment Operators:" << endl;
    cout << "a += 5 * 2; a = " << a << endl;  // Expected: 20
    cout << endl;

    cout << "Conditional Operator (Ternary):" << endl;
    cout << "max = " << max << endl;  // Expected: 20 (since y = 20)
    cout << endl;

    cout << "Comma Operator:" << endl;
    cout << "b = " << b << endl;  // Expected: 20 (since y++ evaluates to 20)
    cout << "x = " << x << ", y = " << y << endl;  // Expected: x = 11, y = 21
    cout << endl;

    return 0;
}
